wicket-core/src/main/java/org/apache/wicket/core/util/string/ComponentRenderer.java

```diff
@@ -16,35 +16,321 @@
+
+			ThreadContext.setRequestCycle(cycle);
+
+			page.get().renderPage();
+
+			return response.getText();
+		});
+	}
+
+	/**
+	 * Run the given runnable inside a valid {@link ThreadContext}.
+	 * 
+	 * @param runnable
+	 *            runnable
+	 */
+	private void inThreadContext(Runnable runnable)
+	{
+		inThreadContext(() -> {
+			runnable.run();
+			return null;
+		});
+	}
+
+	/**
+	 * Get the result from the given supplier inside a valid {@link ThreadContext}.
+	 * 
+	 * @param supplier
+	 *            supplier
+	 * @return result of {@link Supplier#get()}
+	 */
+	private <T> T inThreadContext(Supplier<T> supplier)
+	{
+		ThreadContext oldContext = ThreadContext.detach();
+
+		try
+		{
+			ThreadContext.setApplication(application);
+
+			return supplier.get();
+		}
+		finally
+		{
+
+			ThreadContext.restore(oldContext);
+		}
+	}
+
+	/**
+	 * Create a new request, by default a {@link MockWebRequest}.
+	 */
+	protected WebRequest newWebRequest()
+	{
+		return new MockWebRequest(Url.parse("/"));
+	}
+
+	/**
+	 * Never serialize.
+	 */
+	private static final class NeverSerializer implements ISerializer
+	{
+		@Override
+		public byte[] serialize(Object object)
+		{
+			return null;
+		}
+
+		@Override
+		public Object deserialize(byte[] data)
+		{
+			return null;
+		}
+	}
+
+	/**
+	 * Never share anything.
+	 */
+	private static class NeverSessionStore implements ISessionStore
+	{
+
+		@Override
+		public Serializable getAttribute(Request request, String name)
+		{
+			return null;
+		}
+
+		@Override
+		public List<String> getAttributeNames(Request request)
+		{
+			return null;
+		}
+
+		@Override
+		public void setAttribute(Request request, String name, Serializable value)
+		{
+		}
+
+		@Override
+		public void removeAttribute(Request request, String name)
+		{
+		}
+
+		@Override
+		public void invalidate(Request request)
+		{
+		}
+
+		@Override
+		public String getSessionId(Request request, boolean create)
+		{
+			return null;
+		}
+
+		@Override
+		public Session lookup(Request request)
+		{
+			return null;
+		}
+
+		@Override
+		public void bind(Request request, Session newSession)
+		{
+		}
+
+		@Override
+		public void flushSession(Request request, Session session)
+		{
+		}
+
+		@Override
+		public void destroy()
+		{
+		}
+
+		@Override
+		public void registerUnboundListener(UnboundListener listener)
+		{
+		}
+
+		@Override
+		public void unregisterUnboundListener(UnboundListener listener)
+		{
+		}
+
+		@Override
+		public Set<UnboundListener> getUnboundListener()
+		{
+			return null;
+		}
+
+		@Override
+		public void registerBindListener(BindListener listener)
+		{
+		}
+
+		@Override
+		public void unregisterBindListener(BindListener listener)
+		{
+		}
+
+
+		@Override
+
+		public Set<BindListener> getBindListeners()
+		{
+			return null;
+		}
+	}
+
 	/**
 	 * Collects the html generated by the rendering of a page.
+	 * <p>
+	 * Important note: Must be called on a thread already processing a {@link WebApplication}!
```

Comment:
What does it mean "a thread processing an application" ?
Processing a request is clear, but application is not.

Reply:
`Application.get()` will fail in case current thread doesn't owns Application ....

Why `Application.get()` is being used here, and not `this.application` ?

Reply:
@solomax this is the 'old' static method that reuses whatever application is bound to the thread
@martin-g what about 'a thread bound to a wicket ThreadContext'

Reply:
ups `static`, sorry :(

Reply:
> what about 'a thread bound to a wicket ThreadContext'

Yes, this is more clear, to me at least!